src/model/resources/maintenance.jl

```diff
@@ -0,0 +1,161 @@
+        to begin at any time step during the simulation. Instead this integer describes
+        the cadence of timesteps in which maintenance can begin. Must be at least 1.
+    maint_dur: Number of timesteps that maintenance takes. Must be at least 1.
+    maint_freq_years: 1 is maintenannce every year,
+        2 is maintenance every other year, etc. Must be at least 1.
+    cap: Plant electrical capacity.
+    vcommit: symbol of vCOMMIT-like variable.
+    ecap: symbol of eTotalCap-like variable.
+    integer_operational_unit_committment: whether this plant has integer unit
+        committment for operational variables.
+"""
+function maintenance_constraints!(EP::Model,
+        inputs::Dict,
+        resource_name::AbstractString,
+        suffix::AbstractString,
+        r_id::Int,
+        maint_begin_cadence::Int,
+        maint_dur::Int,
+        maint_freq_years::Int,
+        cap::Float64,
+        vcommit::Symbol,
+        ecap::Symbol,
+        integer_operational_unit_committment::Bool)
+
+    T = 1:inputs["T"]     # Number of time steps (hours)
+    hours_per_subperiod = inputs["hours_per_subperiod"]
+    weights = inputs["omega"]
+
+    y = r_id
+    down_name = maintenance_down_name(resource_name, suffix)
+    shut_name = "vMSHUT_" * resource_name * "_" * suffix
+    down = Symbol(down_name)
+    shut = Symbol(shut_name)
+
+    union!(inputs["MaintenanceDownVariables"], (down,))
+    union!(inputs["MaintenanceShutVariables"], (shut,))
+
+    maintenance_begin_hours = 1:maint_begin_cadence:T[end]
+
+    # create variables
+    vMDOWN = EP[down] = @variable(EP, [t in T], base_name=down_name, lower_bound=0)
+    vMSHUT = EP[shut] = @variable(EP, [t in maintenance_begin_hours],
+                                      base_name=shut_name,
+                                      lower_bound=0)
+
+    if integer_operational_unit_committment
+        set_integer.(vMDOWN)
+        set_integer.(vMSHUT)
+    end
+
+    vcommit = EP[vcommit]
+    ecap = EP[ecap]
+
+    # Maintenance variables are measured in # of plants
+    @constraints(EP, begin
+        [t in T], vMDOWN[t] <= ecap[y] / cap
+        [t in maintenance_begin_hours], vMSHUT[t] <= ecap[y] / cap
+    end)
+
+    # Plant is non-committed during maintenance
+    @constraint(EP, [t in T], ecap[y] / cap - vcommit[y,t] >= vMDOWN[t])
+
+    controlling_hours(t) = controlling_maintenance_start_hours(hours_per_subperiod,
+                                                               t,
+                                                               maint_dur,
+                                                               maintenance_begin_hours)
+    # Plant is down for the required number of hours
+    @constraint(EP, [t in T], vMDOWN[t] == sum(vMSHUT[controlling_hours(t)]))
+
+    # Plant frequire maintenance every (certain number of) year(s)
+    @constraint(EP, sum(vMSHUT[t]*weights[t] for t in maintenance_begin_hours) >=
```

Comment:
how does this work when you are only modeling one year?

Reply:
Yeah, maybe it's pointless to have `weights` here.

In general
* maintenance only works when you are modeling one year
* if a plant requires maintenance every N years then 1/N of the plants require maintenance.